### PR TITLE
Split QgsFileDownloader into separate core/gui classes

### PR DESCRIPTION
**Pull request by nyalldawson**
So that the guts of this class can be used from non-gui code

Refs https://github.com/qgis/QGIS/pull/5384

### CODE REVIEW

**Review by elpaso — APPROVED** [review]
LGTM, I've tested it and there are no regressions.

File: src/core/qgsfiledownloader.h [hunk 1]
```diff
@@ -79,6 +79,9 @@ class GUI_EXPORT QgsFileDownloader : public QObject
      */
     void onDownloadCanceled();
 
+    //! Called to start the download
```
Comment (m-kuhn) [inline src/core/qgsfiledownloader.h]: Can the public slot `onDownloadCanceled` be renamed to `cancelDownload` (imperative)?

File: src/core/qgsfiledownloader.cpp [hunk 2]
```diff
@@ -21,18 +21,17 @@
...
 
-QgsFileDownloader::QgsFileDownloader( const QUrl &url, const QString &outputFileName, bool enableGuiNotifications, const QString &authcfg )
+QgsFileDownloader::QgsFileDownloader(const QUrl &url, const QString &outputFileName, const QString &authcfg , bool delayStart)
   : mUrl( url )
   , mDownloadCanceled( false )
-  , mGuiNotificationsEnabled( enableGuiNotifications )
 {
   mFile.setFileName( outputFileName );
   mAuthCfg = authcfg;
+  if ( !delayStart )
```
Comment (m-kuhn) [inline src/core/qgsfiledownloader.cpp]: indentation?

**Comment by m-kuhn** [comment]
Can you also update the api_break.dox file

Apart from those minor comments, great work!

**Comment by nyalldawson** [comment]
Thanks @m-kuhn, those issues have been addressed.

Looking into this, we have lots of duplicate code which does this same stuff. E.g QgsNetworkContentFetcher, QgsSvgCache::getImageData, QgsGeoNodeRequest, QgsArcGisRestUtils::queryServic, etc. All of these could probably be removed if we added methods to QgsFileDownloader to allow retrieval of the fetched content as a byte array and skip the file writing

**Comment by m-kuhn** [comment]
Yes, download to memory sound useful. Or just add a method that will read file contents to the memory, I guess the bottleneck will always be the net and not the disk. Whatever you prefer.

**Comment by nyalldawson** [comment]
I think you misread my general complaint about "stuff that is wrong and should be fixed" as stating that I'd fix it myself.
.
.
;)

**Comment by m-kuhn** [comment]
We need to start tagging methods and classes with "deprecated, will be removed with qgis 4", it's about time...

**Comment by nyalldawson** [comment]
Yep - probably time as well to grep for all "TODO QGIS 3" and bump to "TODO QGIS 4"...

sigh

**Comment by nyalldawson** [comment]
On the plus side, it means we get to add more sipify magic to translate Q_DECL_DEPRECATED to /Deprecated/

**Pull request merged**